### PR TITLE
Define long MS32 strings.

### PR DESCRIPTION
**Pull request by roconnor-blockstream**

### CODE REVIEW

**Review by apoelstra — APPROVED** [review]

File: MasterSeed32.md [hunk 1]
```diff
@@ -227,7 +227,36 @@ As before, each share must have the same threshold value _t_, the same identifie
 With this set of _t_ MS32 shares, new shares can be derived as discussed above.
 This process generates a fresh master seed, whose value can be retrieved by running the recovery process on any _t_ of these shares.
 
-## Long Strings
+## Long MS32 Strings
```
Comment (robot-dreams) [inline MasterSeed32.md:230]: Nit: is it worth also updating "see long strings" above?

**Review by robot-dreams — COMMENTED** [review]
Nice, this is very helpful.

- Is it worth saying a short sentence about the motivation for introducing a second checksum?
- To avoid ambiguity with the term "secret shares", I renamed the "share data" component of the MS32 string to "payload". How do you feel about using the term "payload" here as well?
- Do we need more detailed instructions (or maybe just a test vector is enough) for the 64-byte case? I'm imagining the "splitting in the middle of a byte" could be a major source of mistakes if not.

File: MasterSeed32.md [hunk 2]
```diff
@@ -227,7 +227,36 @@ As before, each share must have the same threshold value _t_, the same identifie
...
+
+Secrets are limited to 46 bytes when share data is limited to 74 characters.
+However BIP-32 master seeds can be upto 64 bytes in size.
+We define a long MS32 string format to support these longer shares.
+
+* The first 96 characters follow the regular MS32 string format including
+ + A prefix of "ms" or "MS"
+ + A separator, which is always "1".
+ + A 1 charcter theshold parameter
```
[the remark this answers is not in the document]
Comment (roconnor-blockstream) [reply MasterSeed32.md]: Done.

File: MasterSeed32.md [hunk 3]
```diff
@@ -227,7 +227,36 @@ As before, each share must have the same threshold value _t_, the same identifie
...
+ + A 4 character identifier
+ + A 1 character index
+ + A 74 character first share data segment
+ + A 13 character first checksum
+* The remaining characters consiting of
```
[the remark this answers is not in the document]
Comment (roconnor-blockstream) [reply MasterSeed32.md]: Done.

**Comment by roconnor-blockstream** [comment]
> * Is it worth saying a short sentence about the motivation for introducing a second checksum?

I've elaborated on why we are introducing a second checksum.

> * To avoid ambiguity with the term "secret shares", I renamed the "share data" component of the MS32 string to "payload". How do you feel about using the term "payload" here as well?

I've converted to using the term "payload"

> * Do we need more detailed instructions (or maybe just a test vector is enough) for the 64-byte case? I'm imagining the "splitting in the middle of a byte" could be a major source of mistakes if not.

I think the instructions are clear.  We definitely want to add test vectors and also incorrect test vectors.

That said, maybe there is a design that somehow ensures the assembly cannot be done incorrectly?  Seems impossible since the checksum operates on bech32 characters and the byte decoding operates at a lower layer.  

Actually we could make the first payload 30 characters instead of 96.  This is an illegal length decoding a payload which may encourage people to concatenate first.

**Review by apoelstra — APPROVED** [review]
Looks good to me. I think `upto` should be two words everywhere. I'm indifferent about trying to rebalance the lengths to minimize confusion -- though personally I think the existing design is the most intuitive/obvious, even though it does wind up splitting a byte.

**Comment by apoelstra** [comment]
Why is it illegal to decode?

If we are going to do something other than "blocks of 93 characters, with the last one truncated" I'd rather try to make both blocks equal in size.

**Comment by roconnor-blockstream** [comment]
> Why is it illegal to decode?

`Any incomplete group at the end MUST be 4 bits or less`

**Comment by roconnor-blockstream** [comment]
> If we are going to do something other than "blocks of 93 characters, with the last one truncated" I'd rather try to make both blocks equal in size.

BIP-32 technically allows the secret to be any number of bytes between 16 and 64.

**Comment by apoelstra** [comment]
> Any incomplete group at the end MUST be 4 bits or less

Hah! I never noticed that this rule forbade certain lengths of bech32 strings. Very interesting.

Ok, I am torn between the elegance of this, and the fact that it will generalize very poorly to >512-bit data.

**Comment by roconnor-blockstream** [comment]
I think the potential money losing(?) proposition of incorrectly implemented decoders for rare long strings is pretty scary and this  proposal does help mitigate that somewhat.

**Comment by apoelstra** [comment]
I'm doubtful that mistakes here would actually be money-losing: people would decode an incorrect string (in fact, I think if they concatenate after u5->u8 they'll get a *wrong length* string, which would be very visible), find no money on the resulting seed, and hopefully recognize that their decoder was broken and try again with a non-broken one.

I expect that production implementations of this code would have unit tests against exactly this scenario (we should create standard test vectors that exercise it) so even this scenario seems unlikely to me.

I'm also skeptical that somebody who screws up the concatenation logic would correctly implement the "at most 4 bits of padding" logic.

Having said this, there's an argument that the existing "natural" scheme is likely to be implemented without reading the spec very carefully, whereas if you throw curveballs like "30 character prefix which is illegal to decode on its own" into the mix then implementors might keep their eyes open. (But I personally still prefer the existing "natural" design.)

**Comment by roconnor-blockstream** [comment]
Last year we more or less decided to move to using a longer checksum for long secrets. 

Background: 
A 512 bit master seed requires 103 characters to encode it.  We additionally have 6 header characters for threshold value, UID, and share index for a total of 109 characters, not including any checksum.  The existing scheme has length 93 (80 data characters and 13 checksum characters), and our 109 characters exceed the 80 data character limit.

The following spec for a longer checksum for long data has been proposed:

    gen=[23,4,22,5,6,21,23,6,21,25,9,26,25,10,15,1] degree=15 bch_length=1023 bch_dist=9 bch_gen=[23,4,22,5,6,21,23,6,21,25,9,26,25,10,15,1] bch_degree=15 bch_modulus=[25,6,1] bch_c=1019

This is a 15 character checksum.  It has length 1023 and supports up to 1008 data characters which is more than we will ever need (as I recall the math works out so there isn't really anything in between).  We only will use this checksum when the data length exceeds 80 characters as we don't want to burden hand computation with checksums longer than necessary.

The above code has distance 9 and can correct errors in up to 4 characters, same distance as our scheme for short strings.  Note that the error correction _rate_ decreases as data gets longer when the distance remains fixed.

**Comment by apoelstra** [comment]
Sounds good to me. Good observation about the rate. We may want to highlight that for 128- and 256-bit secrets, which use the same code, 128-bit secrets have better error protection for this reason.

**Comment by roconnor-blockstream** [comment]
I've redefined long strings to use a long checksum.

This now requires a thorough review, testing the python code, and we will need to generate at least one more test vector.

**Comment by roconnor-blockstream** [comment]
Hold the presses.   I've run into an issue.

> gen=[23,4,22,5,6,21,23,6,21,25,9,26,25,10,15,1] degree=15

For some reason the generator actually has 16 elements...

**Comment by roconnor-blockstream** [comment]
I've corrected the generators.  The solution was to reverse the list and drop the leading 1.

**Comment by roconnor-blockstream** [comment]
I added a 512 bit example test vector. This is probably ready to be merged now.

**Review by apoelstra — APPROVED** [review]
ACK 10993ab4edda7b8ddd8587ae56f3b7c4bd4ad7a6

I will post a followup PR with flow changes.

**Pull request merged**